frmts/vrt/vrtexpression.cpp

```diff
@@ -0,0 +1,224 @@
+#include "cpl_conv.h"
+#include "cpl_error.h"
+#include "cpl_string.h"
+#include "vrtexpression.h"
+
+#define exprtk_disable_caseinsensitivity
+#define exprtk_disable_rtl_io
+#define exprtk_disable_rtl_io_file
+#define exprtk_disable_rtl_vecops
+#define exprtk_disable_string_capabilities
+#include <exprtk.hpp>
+
+#include <cstdint>
+#include <sstream>
+
+struct vector_access_check final : public exprtk::vector_access_runtime_check
+{
+    bool handle_runtime_violation(violation_context &context) override
+    {
+        auto nElements = (static_cast<std::uint8_t *>(context.end_ptr) -
+                          static_cast<std::uint8_t *>(context.base_ptr)) /
+                         context.type_size;
+        auto nIndexAccessed = (static_cast<std::uint8_t *>(context.access_ptr) -
+                               static_cast<std::uint8_t *>(context.base_ptr)) /
+                              context.type_size;
+
+        std::ostringstream oss;
+        oss << "Attempted to access index " << nIndexAccessed
+            << " in a vector of " << nElements
+            << " elements when evaluating VRT expression.";
+        throw std::runtime_error(oss.str());
+    }
+};
+
+class GDALExpressionEvaluator::Impl
+{
+  public:
+    exprtk::expression<double> m_oExpression{};
+    exprtk::parser<double> m_oParser{};
+    exprtk::symbol_table<double> m_oSymbolTable{};
+    std::string m_osExpression{};
+
+    std::vector<std::pair<std::string, double *>> m_aoVariables{};
+    std::vector<std::pair<std::string, std::vector<double> *>> m_aoVectors{};
+    std::vector<double> m_adfResults{};
+    vector_access_check m_oVectorAccessCheck{};
+
+    bool m_bIsCompiled{false};
+
+    Impl()
+    {
+        using settings_t = std::decay_t<decltype(m_oParser.settings())>;
+
+        m_oParser.register_vector_access_runtime_check(m_oVectorAccessCheck);
+
+        int nMaxVectorSize = std::atoi(
+            CPLGetConfigOption("GDAL_EXPRTK_MAX_VECTOR_SIZE", "100000"));
+
+        if (nMaxVectorSize > 0)
+        {
+            m_oParser.settings().set_max_local_vector_size(nMaxVectorSize);
+        }
+
+        bool bEnableLoops =
+            CPLTestBool(CPLGetConfigOption("GDAL_EXPRTK_ENABLE_LOOPS", "YES"));
```

Comment:
I see in exptrk doc that there is a "Loop Iteration Checks" paragraph with ways to limit the number of iterations or runtime. For a single target pixel computation, it would be crazy that we spend more than  let's say 1 second (and if that threshold is reached, we would error out for following pixels).

Reply:
Here is an example that demonstrates the **_expression evaluation_** timeout functionality:  https://www.partow.net/programming/exprtk/code/exprtk_loop_timeout_rtc.cpp

This is an example that demonstrates both _**expression evaluation and compilation**_ timeout functionality:
https://github.com/google/oss-fuzz/blob/master/projects/exprtk/exprtk_fuzzer.cpp

Reply:
@rouault I added a `GDAL_EXPRTK_TIMEOUT_SECONDS` configuration option in 8e06984befdba8bd8a3c2354fac39e73a26b6a65 . You could exceed the timeout without being caught if you don't use a loop, but hopefully that possibility is excluded by the limits on expression length (`"GDAL_EXPRTK_MAX_EXPRESSION_LENGTH)` and complexity (not currently exposed as a config option).

@ArashPartow exprtk was reporting an iteration _count_ violation when a timeout occurred; see 1d8819d9ea8830b2826df688c2ac9a19dea3e87d